### ISSUE TITLE
SAN option ignored when creating root-ca or intermediate-ca certificate

### ISSUE THREAD

**Issue by frigus02** [smallstep/cli#386]
### Subject of the issue

The `step certificate create` command seems to ignore the `--san` option when creating root-ca or intermediate-ca certificates.

### Your environment

* OS - macOS 10.15.6
* Version - Smallstep CLI/0.15.3 (darwin/amd64), Release Date: 2020-10-22 04:29 UTC

### Steps to reproduce

1. Create a new root-ca certificate, explicitly adding the `--san` argument:

   ```
   step certificate create identity.linkerd.cluster.local ca-new.crt ca-new.key --profile root-ca --no-password --insecure --san identity.linkerd.cluster.local
   ```

1. Inspect the new certificate:

   ```
   step certificate inspect ca-new.crt
   ```

### Expected behaviour

The certificate should have the specified subject alternative name.

### Actual behaviour

The certificate does not have a subject alternative name.

### Additional context

I tried the same with version 1.4.6. Here the created certificate does have a subject alternative name.

**Comment by Pothulapati** [comment]
+1 that san field is not being added in the latest release even when the flag is used.

**Comment by maraino** [comment]
This is on purpose. Root or intermediate certificates are not used for server/client authentication and they do not need to have identities used for authentication.

But it's possible to create a root or intermediate certificate with sans using templates. The default template for a root and intermediate certificate are [here](https://github.com/smallstep/crypto/blob/42c30c598c49b98b32e40a97b60e43588eef885d/x509util/templates.go#L149-L170), and we can adapt them to have SANs, for example, for the root cert, we can add the `"sans"` property:

```
{
	"subject": {{ toJson .Subject }},
	"issuer": {{ toJson .Subject }},
        "sans": {{ toJson .SANs }},
	"keyUsage": ["certSign", "crlSign"],
	"basicConstraints": {
		"isCA": true,
		"maxPathLen": 1
	}
}
```

Storing that in root.tpl, we can can create a root with sans using:
```
step certificate create --template root.tpl --san mariano@smallstep.com "Mariano's Root CA" root_ca.crt root_ca_key
```

**Comment by frigus02** [comment]
Thanks for the quick response. I'm not 100% sure about our use case, yet. But it has to do with go 1.15 validating the SAN:

> x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0

Thanks for the template example. I can confirm that this does add a SAN to the certificate. @Pothulapati, I can update the documentation and change the commands to use custom templates. Or would you rather fix this in a different way?

As a side note, there is still an example in the `step certificate create` command help text, which suggests that using `--san` with `--profile intermediate-ca` should work:

https://github.com/smallstep/cli/blob/d687cbb265bd2df87eb1263d8623c0bfd498129d/command/certificate/create.go#L158-L164

**Comment by maraino** [comment]
@frigus02 the example is a bug, we don't support it right now. I also use Go 1.15 for development, and I haven't seen that error. Are you using an intermediate or root certificate instead of a leaf certificate?

Do you have a snippet of code and a test pki that I can try?

**Comment by Pothulapati** [comment]
@maraino The go 1.15 issue that @frigus02 mentioned is in the linkerd2 repo, as [our documentation](https://linkerd.io/2/tasks/manually-rotating-control-plane-tls-credentials/#generate-a-new-trust-anchor) suggested to use `--san` (whch go 1.15 and hence latest k8s needs with all certs and hence the error when those certs are used with k8s or go 1.15 binaries) to generate certs. This was working till the older version of step and suddenly wasn't :\. We were hoping it was a CLI bug and not something expected. Maybe removing the example should help clarify the confusion. :)

@frigus02 For Linkerd I think we can update the docs to use a templated example. Maybe raise an issue in the Linkerd website repo to have more discussion on this?  Thank you! :)

**Comment by Pothulapati** [comment]
@maraino This is a suggestion. Especially now that with [go 1.15 expects all certs to have the `san` field](https://github.com/golang/go/issues/39568). Why not make it easy for users to do that without having to go through the templating way? I'm thinking a lot of users could run into this.

**Comment by mmalone** [comment]
@Pothulapati we definitely want to make it easy for users to do the right thing. I think the confusion here is around the SAN requirements. There's momentum in the Web PKI community to use SANs instead of a Common Name to identify clients and servers using TLS. So the change you've referenced in go 1.15 makes sense. Common names are weird. They were never well-defined, you can only have one of them, they're not typed (e.g., you can't tell if they're a URI, domain name, email, or just a colloquial name).

However, this is only important for end-entity (aka, _leaf_) certificates.

I think what @maraino was saying above is that, as far as we know, golang 1.15 does not actually require SANs in root or intermediate certificates. It only requires SANs for leaf certificates. Your problem may be that you're attempting to use a root or intermediate certificate as a leaf certificate.

It's quite possible that I'm missing something, but I'm not aware of any requirements or best practices that recommend using SANs in root or intermediate certificates. In fact, it feels sort of weird. It feels like it'd be conflating the namespace for certificate authorities with the namespace for end-entities.

In the Web PKI, for example, root and intermediate certificates have human-readable names like [`"Let's Encrypt Authority X3"`](https://censys.io/certificates/731d3d9cfaa061487a1d71445a42f67df0afca2a6c2d2f98ff7b3ce112b1f568) and [`"GTS CA 101"`](https://censys.io/certificates/95c074e35902a14abd9d19afb6e7f80e669ff8e2363270539d963613f04aaa21). As far as I know, none have SANs. Their names are meant for humans. They aren't domain names, URIs, emails, or anything else that would obviously fit into a SAN field. In fact, the name in a CA certificate is, technically, irrelevant. The important thing for a root certificate is that it's in your trust store. The important thing for an intermediate certificate is that it's signed by a root certificate in your trust store (or signed by another intermediate that chains up to a root certificate in your trust store). The important thing for a _leaf_ certificate is that it chains up to a root in your trust store _and_ has a SAN that matches the name of the thing you're trying to connect to. So names are really only really relevant for leaf certificates.

The problem with putting more stuff in certificates is that other things may depend on that stuff. In this context my concern would be that by putting a domain name in a SAN in a CA certificate you may confuse a TLS client or server into trusting your CA certificate in a TLS handshake. Perhaps that's exactly what you want. But it's definitely _not_ a best practice. You shouldn't be using your CA root certificate as an end-entity certificate. Instead, your CA should issue itself a leaf certificate that it uses for TLS (which can have one or more SANs).

I think I could be convinced to add support for root or intermediate certificate SANs. But I'm not yet. I still don't understand why it's necessary, and I'm worried that it would encourage misuse.

Next step here, I think, would be a strong argument in favor of adding SAN support here and/or some diligence around security considerations.

**Comment by frigus02** [comment]
Thanks a lot for the detailed comment. As you can see the Linkerd team is looking into a different way of verifying root/intermediate certificates.

To me this means you may close the issue or maybe you want to use it to track the removal of the remaining `--san` example.

**Comment by maraino** [comment]
@frigus02 Let's keep it to remove the example.

**Comment by Pothulapati** [comment]
@mmalone @maraino  Thanks for the really useful explanation. We are trying to remove `--san` where we are generating root or intermediate certs and also are removing checks in the code which are verifying root or intermediate certs like leaf certs (and hence the requirement of `SAN` in go 1.15)

**Comment by dopey** [comment]
Hey checking back in. I pushed a doc fix in df41f64.

With regards to adding SANs to the actual certificates -- our current thinking is to wait until someone provides a use case along with explanation for why adding SANs to root/intermediate certs does not pose a security risk. We'll be happy to discuss!

In the mean time I will close this issue. Thanks for bringing the documentation snafu to our attention.
